core/chapters/c12_dictionaries.py

```diff
@@ -0,0 +1,510 @@
+
+The two dictionaries are almost the same, we just swap around the key and value in each pair.
+So to encrypt, we replace `e` with `p`, and to decrypt we change `p` back to `e`.
+
+Note that `'e'` is both a key and a value in `letters`.
+
+Looking up `letters['e']` means that we're asking about `'e'` as a *key*, so it gives `'p'`.
+Remember, we can't use `letters` to ask which key is associated with `'e'` as a *value*.
+But in this case we can use the other dictionary for that: `reverse['e']` gives `'l'`,
+and `letters['l']` gives `'e'` again.
+
+Soon you'll write a function to create a dictionary like `reverse` automatically,
+i.e. `reverse = swap_keys_values(letters)`."""
+
+class DictionaryKeysAndValues(Page):
+    class introducing_keys(Step):
+        """
+Copy this code into the editor, then change `print(quantities)` to `print(quantities.keys())`, and run the whole program.
+
+    __copyable__
+    __no_auto_translate__
+
+    quantities = {'apple': 1, 'cat': 10}
+    print(quantities)
+
+
+        """
+
+        program = """
+        quantities = {'apple': 1, 'cat': 10}
+        print(quantities.keys())
+        """
+
+        def check(self):
+            return "dict_keys(['apple', 'cat'])" in self.result
+
+
+    class keys_are_iterable(VerbatimStep):
+        """
+The `.keys()` method of `dict` does basically what you'd expect. You can iterate over the value it returns
+just like you'd iterate over a list:
+
+    __copyable__
+    __program_indented__
+        """
+
+        def program(self):
+            quantities = {'apple': 1, 'cat': 10}
+            for key in quantities.keys():
+                print(key)
+
+    class keys_are_iterable2(Step):
+        """
+Actually, you don't even need `.keys()`. Iterating directly over a dictionary automatically iterates over its keys.
+Sometimes it's nice to write `.keys()` to make your code more readable, but you don't have to.
+Remove the `.keys()` and run the code again.
+        """
+
+        program = """
+        quantities = {'apple': 1, 'cat': 10}
+        for key in quantities:
+            print(key)
+                """
+
+        def check(self):
+            return "apple\ncat" in self.result
+
+    class cleanup_shopping_cart(ExerciseStep):
+        """
+Now you can use this to modify our function on the previous page to remove the `cart` argument:
+
+    __copyable__
+    __no_auto_translate__
+
+    def total_cost(quantities, prices):
+        result = 0
+        for item in ...:
+            price = prices[item]
+            quantity = quantities[item]
+            result += price * quantity
+        return result
+
+    assert_equal(
+        total_cost(
+            {'dog': 5000000, 'box': 2},
+            {'apple': 2, 'box': 5, 'cat': 100, 'dog': 100},
+        ),
+        500000010,
+    )
+        """
+
+        def solution(self):
+            def total_cost(quantities: Dict[str, str], prices: Dict[str, str]):
```

Comment:
Again, these type hints need correcting, but a `generate_inputs` is needed anyway.